COOKBOOK.md

```diff
@@ -0,0 +1,145 @@
+# The iter-tools Cookbook
+
+Here are some ideas about how iter-tools could simplify common coding tasks!
+
+## Successive functional operations
+If you're already familiar with the advantages of iterators, you should skip this section.
+
+The functional style is very convenient and powerful and javascript. You will frequently find that it is helpful to express transformations on data as a series of calls to methods like Array.map and Array.filter. For example:
+```js
+const users = [
+  [1, {first: 'Ada', last: 'Lovelace', status: 'active'}],
+  null,
+  [3, {status: 'banned'}],
+  [4, {nameless: true, status: 'active'}], // TODO 2/2/1994 remove nameless users
+];
+
+const activeUserNames =
+  users
+    .filter(Boolean) // separate for the purposes of type checking, perhaps
+    .filter(([id, user]) => user.status === 'active')
+    .map(([id, user]) => [user.first, user.last].join(' '))
+    .filter(name => name.length > 0);
+```
+
+In this (possibly slightly contrived) example, the problem is that each of our four functional operations creates a separate array. If our list of users was large to begin with, we're wasting a lot of memory and making unnecessary work for the garbage collector in order to allocate intermediate arrays which are not part of our output anyway. In this regard our function is four times heavier than it needs to be!
+
+Iterators avoid making those intermediate allocations by executing all their specified transformations on an individual item before moving on to the next item.
+
+Using iter-tools, the above example would look like:
+```js
+import { map, filter, compose, toArray } from 'iter-tools';
+
+const activeUserNames = compose(
+    toArray,
+    filter(name => name.length > 0),
+    map(([id, user]) => [user.first, user.last].join(' ')),
+    filter(([id, user]) => user.status === 'active'),
+    filter(Boolean),
+  )(users);
+```
```

Comment:
I would mention that, as of current js engine, for small numbers, arrays is still way faster